src/System.Net.Sockets/src/System/Net/Sockets/MultipleConnectAsync.cs

```diff
@@ -263,9 +263,10 @@ private static Exception AttemptConnection(Socket attemptSocket, SocketAsyncEven
                     NetEventSource.Fail(null, "attemptSocket is null!");
                 }
 
-                if (!attemptSocket.ConnectAsync(args))
+                bool pending = attemptSocket.ConnectAsync(args);
+                if (!pending)
                 {
-                    return new SocketException((int)args.SocketError);
+                    InternalConnectCallback(null, args);
```

Comment:
This fix, and several test fixes, involve handling synchronous completion of ConnectAsync in cases where we didn't handle it properly before.  So these are bugs, but they probably never happened in practice.

What's happening under the covers is that on loopback, an async connect completes asynchronously, but (virtually) instantaneously.  That is, the connect attempt initially is pending, but the epoll notification arrives right away.  Thus our retry-on-epoll-notification kicks in, and the retry succeeds, meaning that from the user's point of view, it looks like a synchronous completion.

This never happened previously because the old locking logic prevented it.  We would never release the queue lock until after we'd done the retry check, so the epoll thread was blocked from updating the queue state and triggering a retry.  That's a problem in and of itself, but it's fixed now.